### DISCUSSION TITLE
Is there a way to obtain related data about DAG dependencies?

### DISCUSSION THREAD

**Discussion by HwiLu** [apache/airflow#34389] [Q&A] [answered]
I want to create a tool that needs to read airflow DAG dependencies data. Then, if any DAG in the dependency graph fails or need to  rerun, it automatically triggers the downstream dependencies for execution，Instead of manually running its dependencies one by one.
My airflow version is 2.5.2.

**Comment by Taragolis** [comment]
[DAG dependencies](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#dag-dependencies) stored as part of Serialised DAG and I don't know any public way to get this data now rather then look at the UI 🤣

**Reply by HwiLu** [reply]
Yes, that's exactly what I wanted to know.

**Reply by HwiLu** [reply]
If there's an API to obtain it, that would work too. I'll continue looking for a solution

**Reply by potiuk** [reply]
> [DAG dependencies](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#dag-dependencies) stored as part of Serialised DAG and I don't know any public way to get this data now rather then look at the UI 🤣

> If there's an API to obtain it, that would work too. I'll continue looking for a solution

Actually yes - there is a public way. See the endpoints here:
https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#tag/DAG

You can query DAG /TASK simplified representation and that includes enough data to reconstruct the dependency graph I believe.

**Reply by Taragolis** [reply]
Yeah, I think in theory it possible reconstruct, my point was about that I believe that we do not return `dag_dependencies` attribute from Serialized DAG in REST API. 

<details>
  <summary>Serialized DAG example_trigger_controller_dag</summary>
  
```json
{
  "dag":{
    "tags":[
      "example"
    ],
    "tasks":[
      {
        "conf":{
          "message":"Hello World"
        },
        "pool":"default_pool",
        "task_id":"test_trigger_dagrun",
        "is_setup":false,
        "ui_color":"#ffefeb",
        "_is_empty":false,
        "_task_type":"TriggerDagRunOperator",
        "ui_fgcolor":"#000",
        "is_teardown":false,
        "_task_module":"airflow.operators.trigger_dagrun",
        "template_ext":[
          
        ],
        "trigger_dag_id":"example_trigger_target_dag",
        "template_fields":[
          "trigger_dag_id",
          "trigger_run_id",
          "execution_date",
          "conf",
          "wait_for_completion"
        ],
        "downstream_task_ids":[
          
        ],
        "wait_for_completion":true,
        "_operator_extra_links":[
          {
            "airflow.operators.trigger_dagrun.TriggerDagRunLink":{
              
            }
          }
        ],
        "on_failure_fail_dagrun":false,
        "template_fields_renderers":{
          "conf":"py"
        }
      }
    ],
    "params":{
      
    },
    "_dag_id":"example_trigger_controller_dag",
    "catchup":false,
    "fileloc":"/files/dags/example_trigger_controller_dag.py",
    "timezone":"UTC",
    "edge_info":{
      
    },
    "timetable":{
      "__var":{
        
      },
      "__type":"airflow.timetables.simple.OnceTimetable"
    },
    "start_date":1609459200.0,
    "_task_group":{
      "tooltip":"",
      "children":{
        "test_trigger_dagrun":[
          "operator",
          "test_trigger_dagrun"
        ]
      },
      "ui_color":"CornflowerBlue",
      "_group_id":null,
      "ui_fgcolor":"#000",
      "prefix_group_id":true,
      "upstream_task_ids":[
        
      ],
      "upstream_group_ids":[
        
      ],
      "downstream_task_ids":[
        
      ],
      "downstream_group_ids":[
        
      ]
    },
    "dag_dependencies":[
      {
        "source":"example_trigger_controller_dag",
        "target":"example_trigger_target_dag",
        "dependency_id":"test_trigger_dagrun",
        "dependency_type":"trigger"
      }
    ],
    "dataset_triggers":[
      
    ],
    "_processor_dags_folder":"/files/dags"
  },
  "__version":1
}
```
  
</details>

**Reply by HwiLu** [reply]
I am sorry,but  may I ask how you obtained such detailed information? The data I receive from the API is quite basic. which is your  API interface？  Is it  :
```
curl -X 'GET' \
  'http://192.168.xxx.xx:8080/api/v1/dags/example_trigger_controller_dag/details' \
  -H 'accept: application/json'
```
what I get like below:
```
{
  "catchup": false,
  "concurrency": 16,
  "dag_id": "example_trigger_controller_dag",
  "dag_run_timeout": null,
  "default_view": "grid",
  "description": null,
  "doc_md": null,
  "end_date": null,
  "file_token": ".eJw9yTEOgCAMQNG7uGsHF29DKlQkVtpARb29TG7_5Q8gauAlEPC1365RbsBpBX1tlzxPC9RkNCr6AyNVwFQ2lhvowVOZXMBYf1hJMVJxXrIVYe7Z_6Tv8AH5Oiet.AJ7Rzh4bu34BV-2bSf4Vk4V4SDo",
  "fileloc": "/opt/code/luhw_venv/lib/python3.8/site-packages/airflow/example_dags/example_trigger_controller_dag.py",
  "is_active": true,
  "is_paused": true,
  "is_paused_upon_creation": null,
  "is_subdag": false,
  "last_parsed": "2023-09-18T06:11:37.847941+00:00",
  "max_active_runs": 16,
  "max_active_tasks": 16,
  "orientation": "LR",
  "owners": [
    "airflow"
  ],
  "params": {},
  "pickle_id": null,
  "render_template_as_native_obj": false,
  "schedule_interval": {
    "__type": "CronExpression",
    "value": "@once"
  },
  "start_date": "2021-01-01T00:00:00+00:00",
  "tags": [
    {
      "name": "example"
    }
  ],
  "timezone": "Timezone('UTC')"
}
```
:relieved:

**Reply by Taragolis** [reply, marked as answer]
> I am sorry,but may I ask how you obtained such detailed information? The data I receive from the API is quite basic. which is your API interface

I don't use any API for get this, it just example what stored into the `serialized_dag` table and also I pointed that no all data might return by REST API. The reason for that usual quite simple: It is not implemented yet, because no one found useful or don't want invest their time to do this.

But Airflow itself it is OSS projects, everyone could propose changes and make this as part of the REST API
- [Contributor's Quick Start](https://github.com/apache/airflow/blob/main/CONTRIBUTORS_QUICK_START.rst#contributors-quick-start)
- [Breeze](https://github.com/apache/airflow/blob/main/BREEZE.rst)
- [CONTRIBUTING.rst](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst)

**Reply by potiuk** [reply]
Yes. As @Taragolis  mentioned the "bulk" information from serialized dags is not available/documented and could be (possibly) contributed) but tou can also use "task simplified representation" https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_task (includes downstream tasks)  - so you can reconstruct the dependencies from there.

**Reply by HwiLu** [reply]
Thank you for  your time
